### PR TITLE
Set header and footer skip line count for CSV tables

### LINKED ISSUE #1085: Allow to set header and footer skip line count for CSV tables

### PR DESCRIPTION
**Pull request by Praveen2112**
Fixes #1085 . Here we replace `text_skip_header_line_count` and `text_skip_footer_line_count` to `skip_header_line_count` and `skip_footer_line_count` which would be common for both csv and text file.

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 1]
```diff
@@ -513,14 +513,14 @@ private ConnectorTableMetadata doGetTableMetadata(SchemaTableName tableName)
             properties.put(AVRO_SCHEMA_URL, avroSchemaUrl);
         }
 
-        // Textfile specific property
-        String textSkipHeaderCount = table.get().getParameters().get(TEXT_SKIP_HEADER_COUNT_KEY);
+        // Textfile and CSV specific property
+        String textSkipHeaderCount = table.get().getParameters().get(SKIP_HEADER_COUNT_KEY);
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: rename variable to `skipHeaderCount`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 2]
```diff
@@ -513,14 +513,14 @@ private ConnectorTableMetadata doGetTableMetadata(SchemaTableName tableName)
...
         if (textSkipHeaderCount != null) {
-            properties.put(TEXTFILE_SKIP_HEADER_LINE_COUNT, Integer.valueOf(textSkipHeaderCount));
+            properties.put(SKIP_HEADER_LINE_COUNT, Integer.valueOf(textSkipHeaderCount));
         }
-        String textSkipFooterCount = table.get().getParameters().get(TEXT_SKIP_FOOTER_COUNT_KEY);
+        String textSkipFooterCount = table.get().getParameters().get(SKIP_FOOTER_COUNT_KEY);
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: same here

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 3]
```diff
@@ -781,24 +781,25 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
...
 
-        // Textfile specific properties
-        getTextHeaderSkipCount(tableMetadata.getProperties()).ifPresent(headerSkipCount -> {
+        // Textfile and CSV specific properties
+        ImmutableSet<HiveStorageFormat> expectedFormats = ImmutableSet.of(HiveStorageFormat.TEXTFILE, HiveStorageFormat.CSV);
+        getHeaderSkipCount(tableMetadata.getProperties()).ifPresent(headerSkipCount -> {
             if (headerSkipCount > 0) {
-                checkFormatForProperty(hiveStorageFormat, HiveStorageFormat.TEXTFILE, TEXTFILE_SKIP_HEADER_LINE_COUNT);
-                tableProperties.put(TEXT_SKIP_HEADER_COUNT_KEY, String.valueOf(headerSkipCount));
+                checkFormatForProperty(hiveStorageFormat, expectedFormats, SKIP_HEADER_LINE_COUNT);
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Maybe:
```
checkFormatForProperty(hiveStorageFormat, HiveStorageFormat.TEXTFILE, TEXTFILE_SKIP_HEADER_LINE_COUNT);
checkFormatForProperty(hiveStorageFormat, HiveStorageFormat.CSV, TEXTFILE_SKIP_HEADER_LINE_COUNT);
```?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 4]
```diff
@@ -781,24 +781,25 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
             tableProperties.put(AVRO_SCHEMA_URL_KEY, validateAndNormalizeAvroSchemaUrl(avroSchemaUrl, hdfsContext));
         }
 
-        // Textfile specific properties
-        getTextHeaderSkipCount(tableMetadata.getProperties()).ifPresent(headerSkipCount -> {
+        // Textfile and CSV specific properties
+        ImmutableSet<HiveStorageFormat> expectedFormats = ImmutableSet.of(HiveStorageFormat.TEXTFILE, HiveStorageFormat.CSV);
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: `expectedFormats` -> `csvAndTextFile`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 5]
```diff
@@ -1333,12 +1341,12 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
                 .collect(toList());
 
         HiveStorageFormat tableStorageFormat = extractHiveStorageFormat(table);
-        if (tableStorageFormat == HiveStorageFormat.TEXTFILE) {
-            if (table.getParameters().containsKey(TEXT_SKIP_HEADER_COUNT_KEY)) {
-                throw new PrestoException(NOT_SUPPORTED, format("Inserting into Hive table with %s property not supported", TEXT_SKIP_HEADER_COUNT_KEY));
+        if (tableStorageFormat == HiveStorageFormat.TEXTFILE || tableStorageFormat == HiveStorageFormat.CSV) {
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: I wonder if this wrapping `if` block is needed here.

File: presto-product-tests/src/main/resources/csv/data.csv [hunk 6]
```diff
@@ -0,0 +1,3 @@
+header1,header2,header3
+value1,value2,value3
+footer1,footer2,footer3
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/csv/data.csv]: no new line at the end

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 7]
```diff
@@ -164,6 +192,87 @@ private void testCreatePartitionedCsvTableAs(String tableName, String additional
...
+                " (col1 int) " +
+                "ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.OpenCSVSerde' " +
+                "STORED AS TEXTFILE " +
+                "TBLPROPERTIES ('skip.header.line.count'='1')");
+        assertThatThrownBy(() -> onPresto().executeQuery("INSERT INTO test_csvfile_skip_header VALUES (1)"))
+                .hasMessageMatching(".* Inserting into Hive table with skip.header.line.count property not supported");
+        onHive().executeQuery("DROP TABLE test_csvfile_skip_header");
+
+        onHive().executeQuery("DROP TABLE IF EXISTS test_csvfile_skip_footer");
+        onHive().executeQuery("" +
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]: can you divide this test into two tests

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 8]
```diff
@@ -164,6 +192,87 @@ private void testCreatePartitionedCsvTableAs(String tableName, String additional
         query("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testCreateTextFileSkipHeaderFooter()
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]: can you please migrate this test to convention tests. See:
```
cox@Grzegorzs-MacBook-Pro:~/presto/presto-product-tests$ find src -name csv*
src/main/resources/sql-tests/datasets/csv_table.ddl
src/main/resources/sql-tests/datasets/csv_table_with_custom_parameters.ddl
src/main/resources/sql-tests/datasets/csv_table_with_custom_parameters.data
src/main/resources/sql-tests/datasets/csv_table.data
src/main/resources/sql-tests/testcases/csv_table_with_custom_parameters.sql
src/main/resources/sql-tests/testcases/csv_table.sql
```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -2179,8 +2179,8 @@ public void testCreateTableWithHeaderAndFooter()
                         ")\n" +
                         "WITH (\n" +
                         "   format = 'TEXTFILE',\n" +
-                        "   textfile_skip_footer_line_count = 1,\n" +
-                        "   textfile_skip_header_line_count = 1\n" +
+                        "   skip_footer_line_count = 1,\n" +
+                        "   skip_header_line_count = 1\n" +
```
Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: can you please add tests for csv with `skip_header_line_count` and `skip_footer_line_count` here?

**Comment by martint** [comment]
@Praveen2112, any updates on this?

**Comment by Praveen2112** [comment]
@martint Yes on the final leg of testing...

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 3 above)
Comment (Praveen2112) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: But a csv format would fail in first statement and similarly for text format.

**Comment by Praveen2112** [comment]
@kokosing , @martint Have applied the comments

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 10]
```diff
@@ -155,20 +155,20 @@
 import static io.prestosql.plugin.hive.HiveTableProperties.ORC_BLOOM_FILTER_COLUMNS;
 import static io.prestosql.plugin.hive.HiveTableProperties.ORC_BLOOM_FILTER_FPP;
 import static io.prestosql.plugin.hive.HiveTableProperties.PARTITIONED_BY_PROPERTY;
+import static io.prestosql.plugin.hive.HiveTableProperties.SKIP_FOOTER_LINE_COUNT;
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: you can do this rename in separate commit, so it will be much easier to review then. In general it is a good practice to have automatic refactors wrapped as separate commits.

File: presto-product-tests/src/main/resources/csv/data.csv [hunk 11]
```diff
@@ -0,0 +1,3 @@
+header1,header2,header3
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/csv/data.csv]: this is very small data test... maybe let's inline this into the java file?

File: presto-product-tests/src/main/resources/csv/data.csv (hunk 6 above)
Comment (kokosing) [inline presto-product-tests/src/main/resources/csv/data.csv]: footer and header does not need to have same number of columns. It is a bit misleading because it looks like data.

```
#### HEADER #######
value1,value2,value3
#### FOOTER #######
```
?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 12]
```diff
@@ -124,6 +126,98 @@ public void testNoPathFilter()
...
+                ImmutableList.of(locatedFileStatus(new Path(SAMPLE_PATH), new DataSize(2.0, GIGABYTE).toBytes())),
+                TupleDomain.all(),
+                Optional.empty(),
+                table,
+                Optional.empty());
+
+        HiveSplitSource hiveSplitSource = hiveSplitSource(backgroundHiveSplitLoader);
+        backgroundHiveSplitLoader.start(hiveSplitSource);
+
+        assertEquals(drainSplits(hiveSplitSource).size(), 33);
```
Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: ❤️ 

Maybe  we can now port tests that were using ./src/main/resources/sql-tests/datasets/table_with_header_and_footer.data-generator?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableProperties.java [hunk 13]
```diff
@@ -129,8 +129,8 @@ public HiveTableProperties(TypeManager typeManager, HiveConfig config)
                         false),
                 integerProperty(BUCKET_COUNT_PROPERTY, "Number of buckets", 0, false),
                 stringProperty(AVRO_SCHEMA_URL, "URI pointing to Avro schema for the table", null, false),
-                integerProperty(TEXTFILE_SKIP_HEADER_LINE_COUNT, "Number of header lines", null, false),
-                integerProperty(TEXTFILE_SKIP_FOOTER_LINE_COUNT, "Number of footer lines", null, false),
+                integerProperty(SKIP_HEADER_LINE_COUNT, "Number of header lines", null, false),
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableProperties.java]: @electrum @findepi do you guys like his rename of `textfile_skip_header_line_count` to `skip_footer_line_count` hive table property?

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 14]
```diff
@@ -13,24 +13,76 @@
...
     private static final String TPCH_SCHEMA = "tiny";
+    private static final String HEADER = "###HEADER###\n";
+    private static final String FOOTER = "\n###FOOTER###";
+    private static final String DATA = "value1,value2,value3";
+
+
+    @Inject
+    private HdfsClient hdfsClient;
+
+    @BeforeTestWithContext
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]: This executes before each test.

The best would be to use test requirements here. So you could define `HiveTableDefinition` that would automatically handle data upload. See `io.prestosql.tests.hive.TestHivePartitionsTable#getRequirements` and usage of `io.prestosql.tempto.fulfillment.table.hive.InlineDataSource#createStringDataSource`

That way you don't need to clean this up, because tempto will take care of this.

Also IIRC tables that are defined in `resources/sql-tests/datasets/` should be also available here via `ImmutableTablesState` so you would need to setup anything here.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 15]
```diff
@@ -164,6 +216,103 @@ private void testCreatePartitionedCsvTableAs(String tableName, String additional
         query("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testCreateCsvFileSkipHeader()
+    {
+        onHive().executeQuery("DROP TABLE IF EXISTS test_create_csvfile_skip_header");
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]: why on Hive? let's use Presto whenever possible

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 16]
```diff
@@ -164,6 +216,103 @@ private void testCreatePartitionedCsvTableAs(String tableName, String additional
...
+                "   csv_separator= ','," +
+                "   skip_header_line_count = 1, " +
+                "   skip_footer_line_count = 1 " +
+                ")");
+        assertThat(query("SELECT * FROM test_create_csvfile_skip_header_footer")).containsExactly(row("value1", "value2", "value3"));
+        onHive().executeQuery("DROP TABLE test_create_csvfile_skip_header_footer");
+    }
+
+    @Test
+    public void testInsertCsvSkipHeader()
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]: can you merge test inserts with create table tests. Tempto tests are a bit heavy.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 17]
```diff
@@ -164,6 +216,103 @@ private void testCreatePartitionedCsvTableAs(String tableName, String additional
...
+    public void testCreateCsvFileSkipHeader()
+    {
+        onHive().executeQuery("DROP TABLE IF EXISTS test_create_csvfile_skip_header");
+        onPresto().executeQuery("" +
+                "CREATE TABLE test_create_csvfile_skip_header" +
+                " (column1 varchar, column2 varchar, column3 varchar) " +
+                "WITH ( " +
+                "   format = 'CSV', " +
+                "   external_location = 'hdfs://hadoop-master:9000/user/hive/warehouse/TestCsvFileHiveTable/csv_table_with_header', " +
+                "   csv_separator= ','," +
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]: can you test different separator like `%`? So we could see that it works with different separators.

File: presto-product-tests/src/main/resources/sql-tests/datasets/csv_with_header_footer.ddl [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+CREATE %EXTERNAL% TABLE %NAME%
+(
+	c_bigint BIGINT,
+	c_varchar VARCHAR(255))
+ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.OpenCSVSerde'
+STORED AS TEXTFILE
+LOCATION '%LOCATION%'
+TBLPROPERTIES (
+    'skip.header.line.count' = '1',
+    'skip.footer.line.count' = '1')
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/sql-tests/datasets/csv_with_header_footer.ddl]: I don't see tests that are using these tables. See:
```
./presto-product-tests/src/main/resources/sql-tests/testcases/csv_table_with_custom_parameters.sql
./presto-product-tests/src/main/resources/sql-tests/testcases/csv_table.sql
```

If you provide convention tests like the ones mentioned above then you can remove all select tests from `TestCsv` and leave there only tests for inserts. Or even you could remove tests for insert from there because that could be easily tested in `TestHiveIntegrationSmokeTest`

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 19]
```diff
@@ -164,6 +172,24 @@ private void testCreatePartitionedCsvTableAs(String tableName, String additional
         query("DROP TABLE " + tableName);
     }
 
+    @Test
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]:
```
@Test(groups = {STORAGE_FORMATS})
```

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 20]
```diff
@@ -164,6 +172,24 @@ private void testCreatePartitionedCsvTableAs(String tableName, String additional
...
+    public void testCreateCsvFileSkipHeader()
+    {
+        assertThat(query("SELECT * FROM csv_with_header")).containsOnly(row("10", "value2"));
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java]: This could be easily refactored to tempto convention tests.

Have you seen:./src/main/resources/sql-tests/testcases/tables_with_header_and_footer.sql?

This way it is able to change tests and then execute them without recompiling.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 21]
```diff
@@ -124,6 +126,98 @@ public void testNoPathFilter()
...
+                ImmutableList.of(locatedFileStatus(new Path(SAMPLE_PATH), new DataSize(2.0, GIGABYTE).toBytes())),
+                TupleDomain.all(),
+                Optional.empty(),
+                table,
+                Optional.empty());
+
+        HiveSplitSource hiveSplitSource = hiveSplitSource(backgroundHiveSplitLoader);
+        backgroundHiveSplitLoader.start(hiveSplitSource);
+
+        assertEquals(drainSplits(hiveSplitSource).size(), 1);
```
Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: can you please extract method like:
```
assertSplitsCount(CSV, ImmutableMap.of("skip.header.line.count", "1", "skip.footer.line.count", "1"), 1 /** splits count */);
```

I see that all the tests you added looked very similar.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 22]
```diff
@@ -49,6 +49,7 @@
 import java.util.ArrayList;
```
Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: I would merge this commit with the previous one.

File: presto-product-tests/src/main/resources/sql-tests/testcases/csv_tables_with_header_and_footer.sql [hunk 23]
```diff
@@ -0,0 +1,13 @@
+-- database: presto; tables: csv_with_header, csv_with_footer, csv_with_header_and_footer; groups: hive, hive_file_header;
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/sql-tests/testcases/csv_tables_with_header_and_footer.sql]: use `groups: storage_formats`
